### PR TITLE
[fix][broker] Avoid using `ArrayList` in the concurrent environment.

### PR DESCRIPTION
**Pull request by mattisonchao**
### Motivation

Relate issue: #16907

https://github.com/apache/pulsar/blob/3f109d6cb7f128a5f8331e33477fd69c3c26d531/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java#L1375-L1442

We have to avoid using `ArrayList` in the concurrent environment.

### Modifications

- Use `synchronizedList` to instead of `ArrayList`

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

- [x] `doc-not-needed` 
(Please explain why)

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -1347,7 +1347,7 @@ public Set<? extends Position> asyncReplayEntries(final Set<? extends Position>
     @Override
     public Set<? extends Position> asyncReplayEntries(Set<? extends Position> positions,
             ReadEntriesCallback callback, Object ctx, boolean sortEntries) {
-        List<Entry> entries = Lists.newArrayListWithExpectedSize(positions.size());
+        List<Entry> entries = Collections.synchronizedList(new ArrayList<>(positions.size()));
```
Comment (shibd) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:1350]: Why should local variables care about concurrency?

Comment (mattisonchao) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:1350]: We will invoke `entries.add` method at callback in different thread.

**Review by tisonkun — COMMENTED** [review]
How can this patch fix the NPE issue reported originally?

BTW, you don't have to sync in `forEach` because `compareAndSet` will return `true` only once.

**Comment by tisonkun** [comment]
I think the issue here can be that the caller calls `readEntryComplete` with `null` parameter.

IIRC bookkeeper handles callbacks one by one so it may not in a concurrent context.

**Comment by mattisonchao** [comment]
Sorry, I missed the `synchronized` in the callback.

**Pull request closed without merging**